### PR TITLE
Add pre-commit + isort

### LINKED ISSUE #90: Adding pre-commit to the library
[pre-commit](https://pre-commit.com/) is a library for managing pre-commit git hooks. Since we already have `black`, `mypy`, `flake8`, etc., it's just a matter of installing and configuring pre-commit to make sure all linters/type checkers are run before committing. We can also add many other useful hooks, e.g. any of the [flake8 extensions](https://github.com/DmytroLitvinov/awesome-flake8-extensions), so that we have more consistent style and minimize the unimportant discussions in code reviews.

I use pre-commit in my local environment, not the least because I sometimes forget to run one of the hooks manually. What do you think of adding this to the library?

### PR DESCRIPTION
**Pull request by leonlan**
Closes #90.
- Add pre-commit to poetry
- Add pre-commit configuration with current dev dependencies
- Add isort as new dependency
- Remove linters/type checkers from poetry dev dependencies

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    rev: 4.0.1
+    hooks:
+    -   id: flake8
+
+-   repo: https://github.com/pycqa/isort
+    rev: 5.10.1
+    hooks:
+      - id: isort
+        name: isort (python)
+        args: [--profile=black, --line-length=79]
```
Comment (N-Wouda) [inline .pre-commit-config.yaml:25]: Can we also add the `case_sensitive` flag here?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+-   repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v3.2.0
+    hooks:
+    -   id: debug-statements
+
+-   repo: https://github.com/psf/black
+    rev: 22.3.0
+    hooks:
+    -   id: black
+        args: [--line-length=79]
```
Comment (N-Wouda) [inline .pre-commit-config.yaml:13]: Question (here, and for other repos below): do these tools not just take settings from `pyproject.toml` already? When I run `black`, I type `poetry run black .`, and then black takes settings from the pyproject file. Is that different when running from precommit?

**Comment by N-Wouda** [comment]
If we move all dev stuff into pre-commit, can we then get rid of the dev dependencies in `pyproject.toml`? I think so, if we update the CI to install and run the pre-commit thing as well?

File: .pre-commit-config.yaml (hunk 2 above)
Comment (leonlan) [reply .pre-commit-config.yaml:13]: Yes this works! I thought the pre-commit did not access it, but I just verified that putting the black/isort/mypy configs in `pyproject.toml` still works.

**Comment by N-Wouda** [comment]
>If we move all dev stuff into pre-commit, can we then get rid of the dev dependencies in pyproject.toml?

Then we can probably also get rid of a lot of the settings in pyproject.toml.

**Comment by leonlan** [comment]
> If we move all dev stuff into pre-commit, can we then get rid of the dev dependencies in `pyproject.toml`? I think so, if we update the CI to install and run the pre-commit thing as well?

I read somewhere that testing is not recommended to be part of pre-commit. I guess that's mainly in very large codebases, which have much longer tests than we do have here. I think it's fine if we do add pytest to pre-commit.

**Comment by N-Wouda** [comment]
>I read somewhere that testing is not recommended to be part of pre-commit. I guess that's mainly in very large codebases, which have much longer tests than we do have here. I think it's fine if we do add pytest to pre-commit.

I meant just the linting and static type checker stuff. The test and coverage stuff can stay, I think.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+-   repo: https://github.com/pycqa/flake8
+    rev: 4.0.1
+    hooks:
+    -   id: flake8
```
Comment (N-Wouda) [inline .pre-commit-config.yaml:18]: Can we also move the args for flake8 in `setup.cfg` here?

Comment (N-Wouda) [reply .pre-commit-config.yaml:18]: (then I'll stop nagging, I promise 😆)

Comment (leonlan) [reply .pre-commit-config.yaml:18]: No problem. There's a lot of things I don't know about, so keep on nagging if you want :-)

**Review by N-Wouda — APPROVED** [review]

File: .github/workflows/alns.yml [hunk 4]
```diff
@@ -27,15 +27,11 @@ jobs:
...
+          poetry run pre-commit install
       - name: Run pytest
         run: poetry run pytest
-      - name: Run black
-        uses: psf/black@stable
-      - name: Run flake8
-        uses: py-actions/flake8@v2
-      - name: Run mypy
-        run: poetry run mypy alns
-      - uses: codecov/codecov-action@v3
```
Comment (N-Wouda) [inline .github/workflows/alns.yml:38]: Issue: can you add the codecov action back in?

Comment (N-Wouda) [reply .github/workflows/alns.yml:38]: I wasn't sure why everything seemed fine, but the codecov status update never came. Now we know :)

Comment (leonlan) [reply .github/workflows/alns.yml:38]: Oops 😅

**Comment by N-Wouda** [comment]
Much better!

**Pull request merged**